### PR TITLE
[PVR] Convert EPG First Aired from time_t to date string

### PR DESCRIPTION
**Pull request by djp952**
## Description
This PR changes the time_t-based "first aired" EPG_TAG into a W3C date string (YYYY-MM-DD) instead.  This allows for dates prior to 1/1/1970 to be displayed in the Kodi UI "Programme Information" screen.

Given the difficulties for backends that cannot provide first aired as a specific date/time that can be localized and compared within the user's time zone, this PR also adds a new EPG_TAG_FLAG to represent "new".  Previously the firstAired time was being compared to the program start time to determine this.

NOTE: **This is a breaking change that will affect all PVR addons;** I understand that they must all be updated prior to acceptance of this PR.  I will need help contacting the maintainers for PVRs I cannot run here

PVR addons will need to be modified to produce a date string rather than a time_t value to supply First Aired.  Additionally, addons will need to have new logic added to determine if an event should be flagged as "new" on their own.  

The addons will now be in full control over both data elements, whatever date they provide for first aired will be displayed as-is, and the "new" indicator will only appear if the addon specifies that it should.

During database migration, existing time_t values will be converted to string dates as possible, although this information is likely to be overwritten by the PVR after the upgrade has taken place.

I also removed the unused "bNotify" flag from the epgtags table since it needs to be upgraded anyway.

edit: After some additional "live" testing with my PVR addon I found that giving the addon the ability to specify just a year instead of a specific date has merit for events like Movies.  PR was updated to include this logic, and screenshots have been added below.

## Motivation and Context
Many EPG backends are incapable of providing a granular enough time_t value that can be localized for the user's time zone and/or compared with the program start time to determine if it's "new" or not.  It's seemingly impossible to provide a "one size fits all" solution.

Being able to specify dates prior to January 1, 1970 is also nice. 

## How Has This Been Tested?
Tested against master branch, Windows x64 Desktop debug build.  Database upgrade path has been tested as has new database path.  Database persistence has been verified as accurate for the data I have available.

I can generate before/after CSV files of an upgraded Epg12 database for review if desired.

## Screenshots (if appropriate):
Sample image with date prior to 1/1/1970 being supported.
pr

Sample image of an EPG event that specifies a Year but not a First Aired date (I believe this is useful for Movie events, perhaps more):
pr-2

Sample image of a pre-1970 event that specifies a First Aired date but not a year:
pr-3

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [X] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project *
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

* There are numerous instances where I left non-clang-format formatting in place to follow conventions within the file; am happy to change any of them.

### CODE REVIEW

**Comment by lrusak** [comment]
jenkins build this please

**Comment by djp952** [comment]
Note: removed errant spacing in xbmc_epg_types.h causing diff ugliness

**Comment by djp952** [comment]
During additional testing I found that having a fallback to just using EPG_TAG.iYear in the "First Aired" display information (like I did for the PVR Recordings PR) has value for events like Movies, and I added that in.

I believe this gives the PVR addon better control of what the user will see based on the information it has available.  If the PVR wants the user to see a specific date, leave iYear zero and set strFirstAired.  If the PVR wants the user to just see a year, only set iYear.  Will add a couple screen shots to the conversation indicating this.

Sorry about changing it on you twice Jenkins.

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (ksooo) [inline xbmc/pvr/epg/EpgDatabase.cpp]: const std::string

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (ksooo) [inline xbmc/pvr/epg/EpgDatabase.cpp]: const std::string

File: xbmc/pvr/guilib/guiinfo/PVRGUIInfo.cpp
Comment (ksooo) [inline xbmc/pvr/guilib/guiinfo/PVRGUIInfo.cpp]: No need for CPVRItem here (expensive) as you have checked the item with IsEPG previously. Use item->GetEPGInfoTag, please.

**Review by ksooo — COMMENTED** [review]
Minors only. Thanks.

**Comment by ksooo** [comment]
> I understand that they must all be updated prior to acceptance of this PR. I will need help contacting the maintainers for PVRs I cannot run here

 Just open PRs against the respective add-ons repos. Maintainers then will pick it up. If not, we consider this addon dead. That's  how it works.

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (Rechi) [inline xbmc/pvr/epg/EpgDatabase.cpp]: > Control statement keywords have to be separated from opening parentheses by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (Rechi) [inline xbmc/pvr/epg/EpgDatabase.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (djp952) [reply xbmc/pvr/epg/EpgDatabase.cpp]: I did not, great catch thank you.  It will need a secondary UPDATE query after the initial SELECT INTO as the data conversion method differs between the two databases.  Will update.

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (djp952) [reply xbmc/pvr/epg/EpgDatabase.cpp]: This one can't be const as it's subsequently assigned a value if .FirstAired().IsValid().  I can make it inline, but I think it would look uglier?

```
  const std::string sFirstAired(
      (tag.FirstAired().IsValid() ? tag.FirstAired().GetAsW3CDate() : ""));
```

Comment (ksooo) [reply xbmc/pvr/epg/EpgDatabase.cpp]: No, was just an oversight on my end. Sorry.

**Comment by djp952** [comment]
First set of deficiencies have been addressed and force-pushed.  Suggest full re-review of CPVREpgDatabase::UpdateTables() changes to ensure the method to differentiate between MySQL and SQLite is acceptable.

I did test the upgrade path again on both SQLite and MySQL, the results were as expected.  I don't think the SQLite version of the UPDATE is necessarily as clean as it could be, but not being able to JOIN during an UPDATE is limiting.

Thank you both! Please let me know what else I messed up.

**Comment by ksooo** [comment]
Seeing you already opening PRs for addons: hold on, please.

We most probably will finish and merge #16485 before we merge new functional API changes. It might be the case that there never will be an addon API version 6.2.0.

@AlwinEsch fyi.

**Comment by djp952** [comment]
Ceased and desisted.  I misunderstood the directive that associated PRs would need to be in place prior to acceptance of this PR.

Erroneous PRs added to pvr.argustv and pvr.dvblink have been closed with an apology note.

**Review by ksooo — APPROVED** [review]

**Comment by AlwinEsch** [comment]
From my side it can in, only I'll have to do a little work on the rebase, but shit happens :grin:.

It would be nice to do it together with my request, but it can't be said exactly when it will come in. Would then reduce the work on the addons since it was all in one.

**Comment by AlwinEsch** [comment]
Also if used this way I can a maybe at helper to the C++ PVR API Interface to help addons where need still time_t.

e.g.  maybe like this:
```cpp
class PVREPGTag : public CStructHdl<PVREPGTag, EPG_TAG>
{
public:
  ...
  void SetStartTime(time_t startTime) { m_cStructure->startTime = startTime; }
  time_t GetStartTime() const { return m_cStructure->startTime; }
```
to
```cpp
class PVREPGTag : public CStructHdl<PVREPGTag, EPG_TAG>
{
public:
  ...
  void SetStartTime(const std::string& startTime) { m_cStructure->startTime = startTime; }
  void SetStartTime(time_t startTime) { m_cStructure->startTime = ToString...Translate(startTime); }
  std::string GetStartTime() const { return m_cStructure->startTime; }
  time_t GetStartTime() const { return ToTime...Translate(m_cStructure->startTime;) }
```

**Comment by ksooo** [comment]
I guess we don't need take helpers.

**Comment by ksooo** [comment]
Keep it simple + separation of concern. These classes should just be dumb containers for c++ <> c conversion, noting else.

**Comment by djp952** [comment]
Rebased as requested.  Seems fine after rebuild.  Only conflict was with indention in EpgDatabase.cpp.

**Comment by phunkyfish** [comment]
As I'm  touching each PVR addon currently we could bring this change in now? Thoughts @ksooo @AlwinEsch ?

@djp952 do you have an example addon PR ready that I could look at?

**Review by AlwinEsch — APPROVED** [review]
From my side good and really super to have. 
As the other takes longer this is good to bring in.

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by djp952** [comment]
@phunkyfish, I had only gotten as far as doing a couple of the trivial ones, i.e. https://github.com/djp952/pvr.argustv/commit/b7f20cbb38bd0c0835c4415dbd1b7137de04c47a.  I had a list at the time of the ones that needed no changes, trivial changes (like argustv), and the ones that needed thought put into it.  IIRC, pretty much each of the non-trivial ones would end up being a little unique based on their input data.

Happy to go back through all the PVRs and recreate that list, also happy to propose PRs for as many as I can manage to run here.

One thought if time is of the essence might be to omit setting firstAired/strFirstAired from the non-trivial PVRs temporarily (like comment it out), to give time to adjust them properly? Seems kind of lame to even suggest such a thing, but I get the impression that there is a time crunch here and I understand finding support for many of these addons has become difficult.

**Comment by ksooo** [comment]
We have got lots of time. Matrix release date is not yet set, we are not even in alpha state.

How can you say something is only for a "short time" if we have no idea how to find maintainers for the addons? 🤔

**Comment by phunkyfish** [comment]
What I meant is once all the PRs for the addons are ready I would merge the kodi PRs and then each of the addon PRs. As it's a breaking change we don't what a large gap (weeks) between merging on kodi and addons PRs.

**Comment by ksooo** [comment]
For the addons not hosted in kodi-pvr only pvr.octonet is not actively maintained any more. I have commit rights for their repo. Just give me a ping when I supposed to merge something.

That does not mean that all kodi-pvr add-ons have an active maintainer. Only some of them have. 😲

**Comment by phunkyfish** [comment]
@djp952 I also have some of them done locally also. I think you will find in your example you need to set `sFirstAired = nullptr;`. If you use empty string it will set first aired as start of Epoch time.

Let me get what I have ready and pushed and then we can see what's left and work through them.

**Comment by ksooo** [comment]
> If you use empty string it will set first aired as start of Epoch time.

 Shouldn't this be fixed? Imo, empty string should be interpreted as first aired date not available.

File: xbmc/pvr/epg/EpgInfoTag.cpp
Comment (phunkyfish) [inline xbmc/pvr/epg/EpgInfoTag.cpp]: How about `if (data.firstAired && strlen(data.strFirstAired) > 0)`. If someone sets this to empty string then that is also invalid?

Comment (AlwinEsch) [reply xbmc/pvr/epg/EpgInfoTag.cpp]: Yes thats make absolutely sense, with the reworked becomes nowhere a `nullptr` given as string.

**Comment by phunkyfish** [comment]
@emveepee if you would like to prep the change for nextpvr based on this PR that would be great. There is a FLAG for New and firstAired is now a string in the format of of W3C date. 

Will ping you once we are ready for everything to go in. There is another related PR here: https://github.com/xbmc/xbmc/pull/17153

Here are some reference implementations: https://github.com/kodi-pvr/pvr.vuplus/pull/275 and  https://github.com/kodi-pvr/pvr.iptvsimple/pull/339

**Comment by emveepee** [comment]
> 
> 
> @emveepee if you would like to prep the change for nextpvr based on this PR that would be great. There is a FLAG for New and firstAired is now a string in the format of of W3C date.

> Will ping you once we are ready for everything to go in. There is another related PR here: #17153

No problem but I will also wait until @ronie updates his PRs https://github.com/xbmc/xbmc/pull/17488 and https://github.com/xbmc/xbmc/pull/17408.  The slowdown issue https://github.com/xbmc/xbmc/issues/17390#issuecomment-599138972 isn't helping move from test build to test build.

**Comment by phunkyfish** [comment]
Sure, all those PRs will go in together, Ronie’s and these ones.

**Comment by phunkyfish** [comment]
@djp952 I have all the PRs ready. I had to comment out first aired on HDHomeRun, MythTV and WMC. All the rest I have them fixed up correctly. Once you have made the changes to this and the other PR (and we have a green build) we can think about starting to merge.

**Comment by phunkyfish** [comment]
jenkins build this please

File: xbmc/pvr/epg/EpgInfoTag.cpp
Comment (djp952) [reply xbmc/pvr/epg/EpgInfoTag.cpp]: Thanks for the save on this one, gentlemen.  Much better this way, it would have been really annoying for people to set an empty string and end up with 1/1/1970.

**Comment by djp952** [comment]
Updated per commentary (thanks again).  I can make the PR for pvr.hdhomerun if desired, and am capable of setting up and working with both WMC and MythTV with what I have available here.  If there is any problem getting those updated I absolutely volunteer.  I made the mess!

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
Cool, once Jenkins catches up and everything is green we can start.

Here are the current PRs for HDHomeRun, MythTV and WMC if you want to work on the follow ups.

- https://github.com/kodi-pvr/pvr.hdhomerun/pull/91
- https://github.com/janbar/pvr.mythtv/pull/139
- https://github.com/kodi-pvr/pvr.wmc/pull/76

**Comment by phunkyfish** [comment]
I can probably use the same method for mythtv and wmc.

**Comment by phunkyfish** [comment]
@djp952 can you test the mythtv and wmc PRs now that I have used a similar fix?

**Comment by djp952** [comment]
You bet, but it will take a little time to set those PVRs up.  Hopefully I can do one tonight and one tomorrow.  They both look like they use a standard time_t so it will probably be perfect

**Comment by phunkyfish** [comment]
Ya, I think they will be ok too. If Jenkins goes green I may just start 😉

**Comment by phunkyfish** [comment]
@djp952 thanks so much for the contribution!

All the kodi-pvr PRs have been merged and released. From the external ones, mythTV, Octonet, telebpy, waipu and zattoo still need their PRs merged and released.

**Comment by phunkyfish** [comment]
@ksooo FYI ^^^

**Comment by djp952** [comment]
Thank you @phunkyfish and @ksooo, I really appreciate the effort that went into merging this.

**Comment by phunkyfish** [comment]
You're very welcome!

**Pull request merged**
